proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -0,0 +1,393 @@
+    "my_global_root_signature",
+    "MyMissShader"
+};
+```
+
+> Note: HLSL State Objects are out of scope for this proposal, and so support
+> for LocalRootSignature and GlobalRootSignature is not covered in this
+> document.
+
+#### Note on the root signature domain specific language
+
+We have received feedback that the DSL for root signatures is not something that
+every language that targets DirectX would want to adopt. For this reason we need
+to ensure that our solution doesn't unnecessarily tie the non-HLSL parts to it.
+
+### Validation and Diagnostics
+
+As well as validating that the root signature is syntactically correct, the
+compiler must also validate that the shader is compatible with the root. For
+example, it must validate that the root signature binds each register that is
+used by the shader. Note that only resources referenced by the entry point need
+to be bound:
+
+```c++
+StructuredBuffer<float4> a : register(t0);
+StructuredBuffer<float4> b : register(t1);
+
+// valid
+[RootSignature("SRV(t0)")]
+float4 eg1() : SV_TARGET { return a[0]; }
+
+// invalid: b is bound to t1 that is not bound in the root signature.
+[RootSignature("SRV(t0)")]
+float4 eg2() : SV_TARGET { return b[0]; } 
+```
+
+## Proposed solution
+
+### Root Signatures in the AST
+
+A new attribute, `HLSLRootSignatureAttr` (defined in `Attr.td`), is added to
+capture the string defining the root signature. `AdditionalMembers` is used to
+add a member that holds the parsed representation of the root signature.
+
+Parsing of the root signature string happens in Sema, and some validation and
+diagnostics can be produced at this stage. For example:
+
+* is the root signature string syntactically correct?
+* is the specified root signature internally consistent?
+  * is the right type of register used in each parameter / descriptor range?
+* is each register bound only once?
+* see [Validations in Sema](#validations-in-sema) for full list
+
+The in-memory representation is guaranteed to be valid as far as the above
+checks are concerned.
+
+The root signature AST nodes are serialized / deserialized as normal bitcode.
+
+In the root signature DSL, a root signature is made up of a list of "root
+elements". The in-memory datastructures are designed around this concept where
+the RootSignature class is a vector of variants.
+
+Example:
+
+```c++
+RootSignature[
+ "RootFlags(ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT),"
+ "CBV(b0, space=1),"
+ "StaticSampler(s1),"
+ "DescriptorTable("
+ "  SRV(t0, numDescriptors=unbounded),"
+ "  UAV(u5, space=1, numDescriptors=10))"
+]
+```
+
+When parsed will produce a the equivalent of:
+
+```c++
+parsedRootSignature = RootSignature{
+  RootFlags(ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT),
+  RootCBV(0, 1), // register 0, space 1
+  StaticSampler(1, 0), // register 1, space 0
+  DescriptorTable({
+    SRV(0, 0, unbounded), // register 0, space 0, unbounded
+    UAV(5, 1, 10) // register 5, space 10, 10 descriptors
```

Comment:
UAV(5, 1, 10) <--- The second parameter here is indicating space 1.  You mentioned space 10 in your comment.  Am I missing something?